ballerina/common_utils.bal

```diff
@@ -351,3 +352,95 @@ isolated function getInputObjectFieldFormPath((string|int)[] path, string name)
+        variables: variables.toJson()
+    };
+    return graphqlPayload;
+}
+
+isolated function handleHttpClientErrorResponse(http:ClientError httpResponse) returns RequestError {
+    if httpResponse is http:ApplicationResponseError {
+        anydata|error data = httpResponse.detail().get("body").ensureType(anydata);
+        if data is error {
+            return error RequestError("GraphQL Client Error", data);
+        }
+        return error RequestError("GraphQL Client Error", body = data);
+    }
+    return error RequestError("GraphQL Client Error", httpResponse);
+}
+
+isolated function handleGraphqlErrorResponse(map<json> responseMap) returns RequestError|ServerError {
+    ErrorDetail[]|error errors = responseMap.get("errors").cloneWithType(GraphQLErrorArray);
+    if errors is error {
+        return error RequestError("GraphQL Client Error", errors);
+    }
+                
+    if (responseMap.hasKey("data") && !responseMap.hasKey("extensions")) {
+        return error ServerError("GraphQL Server Error", data = responseMap.get("data"), errors = errors);
+    } else if (responseMap.hasKey("extensions") && !responseMap.hasKey("data")) {
+        map<json>? extensionsMap = 
+            (responseMap.get("extensions") is ()) ? () : <map<json>> responseMap.get("extensions");
+        return error ServerError("GraphQL Server Error", errors = errors, extensions = extensionsMap);
+    } else if (responseMap.hasKey("data") && responseMap.hasKey("extensions")) {
+        map<json>? extensionsMap = 
+            (responseMap.get("extensions") is ()) ? () : <map<json>> responseMap.get("extensions") ;
+        return error ServerError("GraphQL Server Error", data = responseMap.get("data"), errors = errors, 
+            extensions = extensionsMap);
+    } else {
+        return error ServerError("GraphQL Server Error", errors = errors);
+    }
+}
+
+isolated function performDataBinding(typedesc<GenericResponse|record{}|json> targetType, map<json> responseMap) 
+                                     returns GenericResponse|record{}|json|RequestError {
+    if targetType is typedesc<GenericResponse> {
+        GenericResponse|error response = responseMap.cloneWithType(targetType);
+        if response is error {
+            return error RequestError("GraphQL Client Error", response);
+        }
+        return response;
+    } else if targetType is typedesc<record{}> {
+        record{}|error response = responseMap.cloneWithType(targetType);
+        if response is error {
+            return error RequestError("GraphQL Client Error", response);
+        }
+        return response;
+    } else if targetType is typedesc<json> {
+        json|error response = responseMap.cloneWithType(targetType);
+        if response is error {
+            return error RequestError("GraphQL Client Error", response);
+        }
+        return response;
+    } else {
+        return error RequestError("GraphQL Client Error, Invalid binding type.");
+    }
```

Comment:
Better yet, we may handle all these cases in the native code, so the back and forth inter-op calls can be avoided. (Not 100% sure if that works)

Reply:
This approach doesn't work. Gets the following error
```
ballerina: Oh no, something really went wrong. Bad. Sad.

We appreciate it if you can report the code that broke Ballerina in
https://github.com/ballerina-platform/ballerina-lang/issues with the
log you get below and your sample code.

We thank you for helping make us better.

[2022-03-30 12:58:48,251] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app')
        at org.wso2.ballerinalang.compiler.util.Unifier.getConstraintTypeIfNotError(Unifier.java:670)
        at org.wso2.ballerinalang.compiler.util.Unifier.getTypeAddingArgIfNotProvided(Unifier.java:648)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:572)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:77)
        at org.wso2.ballerinalang.compiler.semantics.model.types.BParameterizedType.accept(BParameterizedType.java:65)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:439)
        at org.wso2.ballerinalang.compiler.util.Unifier.visit(Unifier.java:77)
        at org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType.accept(BUnionType.java:143)
        at org.wso2.ballerinalang.compiler.util.Unifier.build(Unifier.java:99)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationArgs(TypeChecker.java:6728)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParam(TypeChecker.java:6481)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInvocationParamAndReturnType(TypeChecker.java:6360)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.getLangLibMethod(TypeChecker.java:6353)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkInLangLib(TypeChecker.java:3264)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visitInvocation(TypeChecker.java:2970)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.visit(TypeChecker.java:2944)
        at org.wso2.ballerinalang.compiler.tree.expressions.BLangInvocation.accept(BLangInvocation.java:126)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:380)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.TypeChecker.checkExpr(TypeChecker.java:354)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:4014)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangReturn.accept(BLangReturn.java:48)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1902)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:3691)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangDo.accept(BLangDo.java:69)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:1902)
        at org.wso2.ballerinalang.compiler.tree.statements.BLangBlockStmt.accept(BLangBlockStmt.java:65)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeStmt(SemanticAnalyzer.java:4024)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeBlockStmtFollowingIfWithoutElse(SemanticAnalyzer.java:544)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:519)
        at org.wso2.ballerinalang.compiler.tree.BLangBlockFunctionBody.accept(BLangBlockFunctionBody.java:60)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:500)
        at org.wso2.ballerinalang.compiler.tree.BLangFunction.accept(BLangFunction.java:76)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4057)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeNode(SemanticAnalyzer.java:4028)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyzeDef(SemanticAnalyzer.java:4020)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.visit(SemanticAnalyzer.java:349)
        at org.wso2.ballerinalang.compiler.tree.BLangPackage.accept(BLangPackage.java:167)
        at org.wso2.ballerinalang.compiler.semantics.analyzer.SemanticAnalyzer.analyze(SemanticAnalyzer.java:309)
        at io.ballerina.projects.internal.CompilerPhaseRunner.typeCheck(CompilerPhaseRunner.java:188)
        at io.ballerina.projects.internal.CompilerPhaseRunner.performTypeCheckPhases(CompilerPhaseRunner.java:114)
        at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:436)
        at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
        at io.ballerina.projects.ModuleContext.compile(ModuleContext.java:386)
        at io.ballerina.projects.PackageCompilation.compileModulesInternal(PackageCompilation.java:213)
        at io.ballerina.projects.PackageCompilation.compileModules(PackageCompilation.java:198)
        at io.ballerina.projects.PackageCompilation.compile(PackageCompilation.java:114)
        at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:109)
        at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:233)
        at io.ballerina.projects.Package.getCompilation(Package.java:145)
        at io.ballerina.projects.Package.runCodeGeneratorPlugins(Package.java:250)
        at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:84)
        at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
        at io.ballerina.cli.cmd.TestCommand.execute(TestCommand.java:265)
        at java.base/java.util.Optional.ifPresent(Unknown Source)
        at io.ballerina.cli.launcher.Main.main(Main.java:51)
 
ERROR [graphql:(1:1,1:1)] Compilation failed due to: class org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType cannot be cast to class org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType (org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType and org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType are in unnamed module of loader 'app')
error: compilation contains errors
```

The reason we didn't handle it in the native code is because the `processExecute` function includes a HTTP post call. Handling the HTTP call is more easier in Ballerina code. If we bring this function to native code it would increase the the back and forth inter-op calls.